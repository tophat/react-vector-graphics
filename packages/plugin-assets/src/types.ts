export type PluginAssetsOptions = {
    'assets/fileExt': string
    'assets/globPattern': string
    'assets/nameScheme': string
    'assets/outputPath': string
}
