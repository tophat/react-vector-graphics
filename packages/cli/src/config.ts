import { cosmiconfig } from 'cosmiconfig'

import type { Config } from '@react-vector-graphics/types'

const cache = process.env.NODE_ENV !== 'test'
const explorer = cosmiconfig('rvg', { cache })

export const init = (): Config => ({
    plugins: [],
})
export const find = explorer.search
export const load = explorer.load
