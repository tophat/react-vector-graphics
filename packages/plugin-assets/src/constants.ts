export const OPTIONS = {
    FILE_EXT: 'assets/fileExt',
    GLOB_PATTERN: 'assets/globPattern',
    NAME_SCHEME: 'assets/nameScheme',
    OUTPUT_PATH: 'assets/outputPath',
} as const

export const STATE = {
    COMPONENT_FILES: 'assets/componentFiles',
    COMPONENT_NAME: 'componentName', // @svgr/plugin-jsx
    FILE_PATH: 'filePath', // @svgr/plugin-jsx
} as const
