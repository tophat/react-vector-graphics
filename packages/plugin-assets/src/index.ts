export * from './constants'
export type { PluginAssetsOptions } from './types'
export { run as default } from './main'
