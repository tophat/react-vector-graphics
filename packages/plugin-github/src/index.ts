export * from './constants'
export { run as default } from './main'
export type { PluginGitHubOptions } from './types'
