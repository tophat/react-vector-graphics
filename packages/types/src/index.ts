import type { ConfigPlugin, Config as CoreConfig, State } from '@svgr/core'

export type ReactVectorGraphicsOptions = {
    // API
    // FOLDER_PATH
    // HEAD
    // OWNER
    // REPO
}

export type Config = Omit<CoreConfig, 'plugins'> & {
    options?: Record<string, unknown>
    plugins?: (ConfigPlugin | Plugin)[]
}
/** @deprecated Use Config instead. */
export type Configuration = Config

export type { State } from '@svgr/core'
export type AnyObject = { [key: string]: any }

export type Log = Console['log']

export type PluginParams = {
    code: string
    state: State
}

export type PluginResult = string | PluginParams | string[] | PluginParams[]

export type Logger = { error: Log; info: Log; warn: Log }

export type Plugin = (
    code: string | undefined,
    config: Config,
    state: Omit<State, 'componentName'> & { componentName?: string },
    logger?: Logger,
) => Promise<PluginResult>

export type NamingScheme =
    | 'CONSTANT_CASE'
    | 'camelCase'
    | 'PascalCase'
    | 'snake_case'
    | 'spinal-case'
