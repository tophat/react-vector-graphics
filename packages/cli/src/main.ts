import * as path from 'path'

import { getLogger, default as rvgCore } from '@react-vector-graphics/core'
import yargs from 'yargs'

import * as conf from './config'

import type { Config, Logger } from '@react-vector-graphics/types'
import type { CosmiconfigResult } from 'cosmiconfig/dist/types'

const configResultHandler =
    (config: Config, logger: Logger) =>
    (result: CosmiconfigResult): void => {
        if (result?.filepath && !result.isEmpty) {
            logger.info(`Loaded config from ${result.filepath}`)
            Object.assign(config, result.config)
            const configDir = path.dirname(result.filepath)
            if (configDir !== process.cwd()) {
                logger.info(`Changing current directory to ${configDir}`)
                process.chdir(configDir)
            }
        } else if (result?.isEmpty) {
            logger.info(`Skipping empty config at ${result.filepath}`)
        } else {
            logger.info('No config loaded')
        }
    }

const getConfig = async (config: Config, logger: Logger): Promise<Config> => {
    const currentDir = process.cwd()
    const handleConfig = configResultHandler(config, logger)
    logger.info(`Searching for config in ${currentDir}`)
    const result = await conf.find(currentDir)
    handleConfig(result)

    logger.info('Start getting cli options')
    const { argv } = yargs(process.argv)
        .usage('Usage: $0 --config [config]')
        .option('config', {
            default: result?.filepath,
            describe: 'Config file path',
            type: 'string',
        })
        .demandOption(['config'])

    if (result?.filepath !== argv.config) {
        await conf.load(argv.config).then(handleConfig)
    }
    return config
}

export const run = async (): Promise<void> => {
    const defaultConfig = conf.init()
    const logger = getLogger()
    const config = await getConfig(defaultConfig, logger)
    if (!config.plugins?.length) {
        logger.info('Setting the minimal plugin configuration')
        const assetPlugin = '@react-vector-graphics/plugin-assets'
        config.plugins = [assetPlugin, '@svgr/plugin-jsx', assetPlugin]
    }
    await rvgCore({ config, logger })
}
