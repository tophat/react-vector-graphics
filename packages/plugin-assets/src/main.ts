import * as path from 'path'

import { pathToName } from '@react-vector-graphics/utils'
import * as fs from 'fs-extra'
import * as glob from 'glob'

import { OPTIONS, STATE } from './constants'

import type { PluginAssetsOptions } from './types'
import type {
    Logger,
    NamingScheme,
    Plugin,
    PluginParams,
    State,
} from '@react-vector-graphics/types'

const findAssets = (params: {
    globPattern: string
    nameScheme: NamingScheme
    state: Omit<State, 'componentName'> & { componentName?: string }
}): PluginParams[] => {
    return glob.sync(params.globPattern).map(
        (filePath: string): PluginParams => ({
            code: fs.readFileSync(filePath, { encoding: 'utf-8' }),
            state: Object.assign(
                {
                    [STATE.COMPONENT_NAME]: pathToName(
                        filePath,
                        params.nameScheme,
                    ),
                    [STATE.FILE_PATH]: filePath,
                },
                params.state,
            ) as State,
        }),
    )
}

const writeComponent = (params: {
    assetFile?: string
    code: string
    componentName?: string
    componentFiles: { [fileName: string]: string }
    fileExt?: string | undefined
    logger?: Logger
    outputPath?: string
}): void => {
    if (!params.componentName) {
        return params.logger?.warn(
            `No '${STATE.COMPONENT_NAME}' provided for '${params.assetFile}'.`,
        )
    }
    if (!params.outputPath) {
        return params.logger?.warn(`No '${OPTIONS.OUTPUT_PATH}' provided.`)
    }
    if (!params.fileExt) {
        params.logger?.warn(`No '${OPTIONS.FILE_EXT}' provided.`)
    }

    const componentFiles = Object.entries(params.componentFiles)
    const pathToFolder = path.join(
        params.outputPath,
        componentFiles.length ? params.componentName : '',
    )
    const pathToFile = path.join(
        pathToFolder,
        componentFiles.length ? 'index' : params.componentName,
    )
    const componentFilePath = params.fileExt
        ? `${pathToFile}.${params.fileExt}`
        : pathToFile
    fs.outputFileSync(componentFilePath, params.code)
    for (const [fileName, fileContents] of componentFiles) {
        const filePath = path.join(pathToFolder, fileName)
        fs.outputFileSync(filePath, fileContents)
    }
}

export const run: Plugin = async (code, config, state, logger) => {
    const options: Partial<PluginAssetsOptions> = config.options ?? {}

    if (code) {
        const opts = {
            assetFile: state[STATE.FILE_PATH],
            code,
            // @ts-expect-error componentFiles is added to state
            componentFiles: state[STATE.COMPONENT_FILES] ?? {},
            componentName: state[STATE.COMPONENT_NAME],
            fileExt: options?.[OPTIONS.FILE_EXT],
            logger,
            outputPath: options?.[OPTIONS.OUTPUT_PATH],
        }
        writeComponent(opts)
        return { code, state: state as State }
    } else {
        if (!options?.[OPTIONS.GLOB_PATTERN]) {
            throw new Error('Invariant violation. Missing glob pattern.')
        }
        return findAssets({
            // @ts-expect-error Args need to be pre-validated
            globPattern: options?.[OPTIONS.GLOB_PATTERN],
            nameScheme: options?.[OPTIONS.NAME_SCHEME] as NamingScheme,
            state,
        })
    }
}
