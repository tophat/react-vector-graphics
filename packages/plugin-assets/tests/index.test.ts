import * as path from 'path'

import { fs, vol } from 'memfs'

import { OPTIONS, STATE, default as assetsPlugin } from '../src'

import type { Config, State } from '@react-vector-graphics/types'

describe('plugin-assets', () => {
    const mockSVGContent = '<svg></svg>'
    const mockFilePath = 'example/assets/file3.svg'
    const mockComponentPascalName = 'File3'
    const spyLogWarn = jest.spyOn(console, 'warn')

    beforeEach(() => {
        vol.fromJSON({
            'example/assets/file-1.24x24.icon.svg': mockSVGContent,
            'example/assets/file2.icon.svg': mockSVGContent,
            [mockFilePath]: mockSVGContent,
        })
    })

    afterEach(() => {
        jest.resetAllMocks()
        vol.reset()
    })

    it('finds all files matching pattern', async () => {
        const config: Config = {
            options: {
                [OPTIONS.GLOB_PATTERN]: 'example/assets/*.icon.svg',
            },
            plugins: [],
        }
        expect(await assetsPlugin('', config, {}, console)).toMatchSnapshot()
    })

    it('uses supplied naming scheme', async () => {
        const config: Config = {
            options: {
                [OPTIONS.GLOB_PATTERN]: 'example/**/*.svg',
                [OPTIONS.NAME_SCHEME]: 'snake_case',
            },
            plugins: [],
        }
        expect(await assetsPlugin('', config, {}, console)).toMatchSnapshot()
    })

    it('writes single component file when given code', async () => {
        const fileExt = 'js'
        const outputPath = 'example/components'
        const pathToFile = path.join(
            outputPath,
            `${mockComponentPascalName}.${fileExt}`,
        )
        const state: State = {
            [STATE.FILE_PATH]: mockFilePath,
            [STATE.COMPONENT_NAME]: mockComponentPascalName,
        }
        const config: Config = {
            options: {
                [OPTIONS.FILE_EXT]: fileExt,
                [OPTIONS.OUTPUT_PATH]: outputPath,
            },
            plugins: [],
        }

        expect(fs.existsSync(pathToFile)).toBe(false)
        await assetsPlugin(mockSVGContent, config, state, console)
        expect(fs.readFileSync(pathToFile).toString()).toEqual(mockSVGContent)
        expect(spyLogWarn).not.toHaveBeenCalled()
    })

    it('writes files to component folder when given extra files', async () => {
        const fileExt = 'js'
        const outputPath = 'example/components'
        const mockFileName = `index.${fileExt}`
        const mockExtraFileName = 'README.md'
        const mockExtraFileContent = `# ${mockComponentPascalName}`
        const pathToFolder = path.join(outputPath, mockComponentPascalName)
        const pathToIndexFile = path.join(pathToFolder, mockFileName)
        const pathToExtraFile = path.join(pathToFolder, mockExtraFileName)
        const state: State = {
            [STATE.FILE_PATH]: mockFilePath,
            [STATE.COMPONENT_NAME]: mockComponentPascalName,
            [STATE.COMPONENT_FILES]: {
                [mockExtraFileName]: mockExtraFileContent,
            },
        } as any
        const config: Config = {
            options: {
                [OPTIONS.FILE_EXT]: fileExt,
                [OPTIONS.OUTPUT_PATH]: outputPath,
            },
            plugins: [],
        }

        expect(fs.existsSync(pathToIndexFile)).toBe(false)
        expect(fs.existsSync(pathToExtraFile)).toBe(false)
        await assetsPlugin(mockSVGContent, config, state, console)
        expect(fs.existsSync(pathToIndexFile)).toBe(true)
        expect(fs.readFileSync(pathToExtraFile).toString()).toEqual(
            mockExtraFileContent,
        )
        expect(spyLogWarn).not.toHaveBeenCalled()
    })

    it('warns and continues when fileExt is not provided', async () => {
        const outputPath = 'example/components'
        const pathToFile = path.join(outputPath, mockComponentPascalName)
        const state: State = {
            [STATE.FILE_PATH]: mockFilePath,
            [STATE.COMPONENT_NAME]: mockComponentPascalName,
        }
        const config: Config = {
            options: {
                [OPTIONS.OUTPUT_PATH]: outputPath,
            },
            plugins: [],
        }

        expect(fs.existsSync(pathToFile)).toBe(false)
        await assetsPlugin(mockSVGContent, config, state, console)
        expect(fs.readFileSync(pathToFile).toString()).toEqual(mockSVGContent)
        expect(spyLogWarn).toHaveBeenCalled()
        expect(spyLogWarn.mock.calls[0][0]).toMatchSnapshot()
    })

    it('warns and exits when componentName is not provided', async () => {
        const fileExt = 'js'
        const outputPath = 'example/components'
        const pathToFile = path.join(
            outputPath,
            `${mockComponentPascalName}.${fileExt}`,
        )
        const state: State = {
            [STATE.FILE_PATH]: mockFilePath,
        } as any
        const config: Config = {
            options: {
                [OPTIONS.FILE_EXT]: fileExt,
                [OPTIONS.OUTPUT_PATH]: outputPath,
            },
            plugins: [],
        }

        expect(fs.existsSync(pathToFile)).toBe(false)
        await assetsPlugin(mockSVGContent, config, state, console)
        expect(fs.existsSync(pathToFile)).toBe(false)
        expect(spyLogWarn).toHaveBeenCalled()
        expect(spyLogWarn.mock.calls[0][0]).toMatchSnapshot()
    })

    it('warns and exits when outputPath is not provided', async () => {
        const fileExt = 'js'
        const state: State = {
            [STATE.FILE_PATH]: mockFilePath,
            [STATE.COMPONENT_NAME]: mockComponentPascalName,
        }
        const config: Config = {
            options: {
                [OPTIONS.FILE_EXT]: fileExt,
            },
            plugins: [],
        }

        await assetsPlugin(mockSVGContent, config, state, console)
        expect(spyLogWarn).toHaveBeenCalled()
        expect(spyLogWarn.mock.calls[0][0]).toMatchSnapshot()
    })
})
