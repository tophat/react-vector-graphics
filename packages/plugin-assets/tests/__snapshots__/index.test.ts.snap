// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-assets finds all files matching pattern 1`] = `
[
  {
    "code": "<svg></svg>",
    "state": {
      "componentName": "File124x24Icon",
      "filePath": "example/assets/file-1.24x24.icon.svg",
    },
  },
  {
    "code": "<svg></svg>",
    "state": {
      "componentName": "File2Icon",
      "filePath": "example/assets/file2.icon.svg",
    },
  },
]
`;

exports[`plugin-assets uses supplied naming scheme 1`] = `
[
  {
    "code": "<svg></svg>",
    "state": {
      "componentName": "file_1_24x24_icon",
      "filePath": "example/assets/file-1.24x24.icon.svg",
    },
  },
  {
    "code": "<svg></svg>",
    "state": {
      "componentName": "file2_icon",
      "filePath": "example/assets/file2.icon.svg",
    },
  },
  {
    "code": "<svg></svg>",
    "state": {
      "componentName": "file3",
      "filePath": "example/assets/file3.svg",
    },
  },
]
`;

exports[`plugin-assets warns and continues when fileExt is not provided 1`] = `"No 'assets/fileExt' provided."`;

exports[`plugin-assets warns and exits when componentName is not provided 1`] = `"No 'componentName' provided for 'example/assets/file3.svg'."`;

exports[`plugin-assets warns and exits when outputPath is not provided 1`] = `"No 'assets/outputPath' provided."`;
